### ISSUE TITLE
Would you be interested in a PR adding type hints?

### ISSUE THREAD

**Issue by Recursing** [pdfminer/pdfminer.six#362]
I think type hints would help users and new contributors

**Comment by pietermarsman** [comment]
Yes, that's interesting! As long as it does not break anything; I am not sure if type hints are supported by Python3.4. 

Some additional thoughts: 

- you can start by adding types to the tools (e.g. pdf2txt.py and dumppdf.py), the high_level.py functions and the most used classes (e.g. `PDFDocument`, `TextConverter`, `PDFPageAggregator`, `LAParams`, `PDFResourceManager`, `PDFPageInterpreter` and `PDFPage`). 
- [mypy](http://mypy-lang.org/) is a static type checker that could enforce the correct usage of types
- be sure to read [pep484](https://www.python.org/dev/peps/pep-0484/)

**Comment by Recursing** [comment]
Python 3.4 does not support inline type annotations, so the valid syntax would be annoyingly verbose (and non-standard). But why do you guys even support it? I think it's not that common anymore

**Comment by pietermarsman** [comment]
I also think that very few people use that. It is much easier to upgrade from python 3.4 than it was from e.g. 2.7.

Let me check with other contributors to see what they think.

**Comment by jstockwin** [comment]
@pietermarsman Did you get any response from other contributors? In my view I think it's fine to drop python 3.4 support, and type hints would be nice.

The only drawback is it might be a slight blocker for other people contributing if they don't know how to use them.

@Recursing I appreciate it's been a long time, but we're currently trying to sort through all the open issues. Is this something you'd still be interested in helping with?

**Comment by Recursing** [comment]
> The only drawback is it might be a slight blocker for other people contributing if they don't know how to use them.

They don't need to be mandatory. I think they would definitely help more than harm

> Is this something you'd still be interested in helping with?

Sure

**Comment by pietermarsman** [comment]
> @pietermarsman Did you get any response from other contributors? In my view I think it's fine to drop python 3.4 support, and type hints would be nice.

Not that I remember. 

I like type annotations so I am pro :)

**Comment by jstockwin** [comment]
I also like them so I vote yes

**Comment by pietermarsman** [comment]
When we add this, we should also add mypy to the travis-ci test run. I just ran it locally, and there are only 8 issues. This surprised me a bit, but maybe its because we dont have type annotations yet. With mypy the type annotations add more value because we detect wrong usage of types earlier.

**Comment by jstockwin** [comment]
Agreed. Potentially also worth adding pytype as the two tools check different things

**Comment by pietermarsman** [comment]
I also wanted to add mypy to a work-related project, but after you suggested pytype I did a little research into that. And now I actually prefer pytype over mypy because of two benefits:

- pytype infers types from initialization methods, e.g. if you use `str` it infers the variable is a `str`. Mypy does not infer types and only relies on type annotations so much less coverage. 
- It does type inference for every line, where mypy allows a single variable to have a single type during its entire lifetime. The latter might sound more strict and understandable, but it does not allow to map 'str' over a list or do multiple operations to the same variable. 

See this [presentation](https://docs.google.com/presentation/d/1GYqLeLkknjYaYX2JrMzxX8LGw_rlO-6kTk-VNPVG9gY/edit) (by creator of pytype)

**Comment by jstockwin** [comment]
Yeah, I agree it's good. However, there are some things that mypy catches that pytype won't, so using both is a valid option.

**Comment by pietermarsman** [comment]
Lets leave it up to the one who implements it then :p I've moved this to accepted.

**Comment by djbrown** [comment]
Wouldn't it be best to add the stubs in https://github.com/python/typeshed?
Then the type hints would be completely optional.
I'm currently looking into this with local stubs.

**Comment by pietermarsman** [comment]
Closed by #661
